x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -295,6 +318,21 @@ private void buildThenMaybeCacheRole(RoleKey roleKey, Collection<RoleDescriptor>
         }, listener::onFailure));
     }
 
+    private void getOrBuildRoleForApiKey(Authentication authentication, boolean limitedBy, ActionListener<Role> roleActionListener) {
+        final BytesReference roleDescriptorsBytes = apiKeyService.getRoleDescriptorsBytesForApiKey(authentication, limitedBy);
+        final String roleDescriptorsHash = new String(Hasher.SHA256.hash(new SecureString(roleDescriptorsBytes.toString().toCharArray())));
```

Comment:
I really like `BytesKey`, unfortunately it is not compatible with the current cache and I don't want to add another cache and associated overhead just because of this incompatibility. So I'll prefer to your second suggestion, i.e. `MessageDigest.sha256` where raw bytes are used.

Alternatively, we could potential change `RoleKey` to have either `Set<byte[]> names` and `byte[] source`, so it can accomodate both regular roles and API key roles. But I am not sure whether it really gives us much unless we have a real concern for sha256 collision.

Reply:
I'm happy with the sha256 option. I also like `BytesKey` - it's simple and sensible, but I don't think it's the right fit here unless we change the cache to have a`Set<?>` as the key, or follow your idea of converting a `Set<String>` into a `BytesKey`

Reply:
Cool let's keep the sha256 option then. It works and is simpler.